swift/extractor/infra/log/SwiftLogging.h

```diff
@@ -0,0 +1,164 @@
+class Log {
+ public:
+  using Level = binlog::Severity;
+
+ private:
+  // Output filtered according to a configured log level
+  template <typename Output>
+  struct FilteredOutput {
+    binlog::Severity level;
+    Output output;
+    binlog::EventFilter filter{
+        [this](const binlog::EventSource& src) { return src.severity >= level; }};
+
+    template <typename... Args>
+    FilteredOutput(Level level, Args&&... args)
+        : level{level}, output{std::forward<Args>(args)...} {}
+
+    FilteredOutput& write(const char* buffer, std::streamsize size) {
+      filter.writeAllowed(buffer, size, output);
+      return *this;
+    }
+
+    // if configured as `no_logs`, the output is effectively disabled
+    explicit operator bool() const { return level < Level::no_logs; }
+  };
+
+  using LevelRule = std::pair<std::regex, Level>;
+  using LevelRules = std::vector<LevelRule>;
+
+  static constexpr const char* format = "%u %S [%n] %m (%G:%L)\n";
+
+  binlog::Session session;
+  std::ofstream textFile;
+  FilteredOutput<std::ofstream> binary{Level::no_logs};
+  FilteredOutput<binlog::TextOutputStream> text{Level::info, textFile, format};
+  FilteredOutput<binlog::TextOutputStream> console{Level::warning, std::cerr, format};
+  LevelRules sourceRules;
+  std::string rootName;
+
+  Log() = default;
+
+  static Log& instance() {
+    static Log ret;
+    return ret;
+  }
+
+  friend class Logger;
+  friend binlog::Session;
+
+  Level getLevelForSource(std::string_view name) const;
+  Log& write(const char* buffer, std::streamsize size);
+  static class Logger& logger();
+
+ public:
+  // Configure logging. This consists in
+  // * setting up a default logger with `root` as name
+  // * using environment variable `CODEQL_EXTRACTOR_SWIFT_LOG_DIR` to choose where to dump the log
+  //   file(s). Log files will go to a subdirectory thereof named after `root`
+  // * using environment variable `CODEQL_EXTRACTOR_SWIFT_LOG_LEVELS` to configure levels for
+  //   loggers and outputs. This must have the form of a comma separated `spec:level` list, where
+  //   `spec` is either a glob pattern (made up of alphanumeric, `/`, `*` and `.` characters) for
+  //   matching logger names or one of `out:bin`, `out:text` or `out:console`.
+  //   Output default levels can be seen in the corresponding initializers above. By default, all
+  //   loggers are configured with the lowest output level
+  static void configure(std::string_view root);
+
+  // Flush logs to the designated outputs
+  static void flush();
+};
+
+// This class represent a named domain-specific logger, responsible for pushing logs using the
+// underlying `binlog::SessionWriter` class. This has a configured log level, so that logs on this
+// `Logger` with a level lower than the configured one are no-ops.
+class Logger {
+  binlog::SessionWriter w{Log::instance().session};
+  Log::Level level_{Log::Level::no_logs};
+
+  void setName(std::string name);
```

Comment:
SGTM

Reply:
I've decided for removing the default `codeql::logger` for now, we will see if we really need it. I use a `"main"` logger in the `main` function instead of this default one, built locally over there.